File: static/js/calculator.js
/**
 * Основной модуль калькулятора автокредита
 */

// Глобальные переменные
let earlyPaymentsData = {};
let currentSchedule = [];
let currentParameters = {};
let isInstallmentMode = false;
let termUnit = 'months'; // 'months' или 'years'

// Инициализация при загрузке страницы
document.addEventListener('DOMContentLoaded', function() {
    initializeApp();
});

/**
 * Инициализация приложения
 */
function initializeApp() {
    // Устанавливаем текущую дату по умолчанию
    const today = new Date().toISOString().split('T')[0];
    document.getElementById('startDate').value = today;
    
    // Обработчики событий
    setupEventListeners();
    
    // Инициализация форматирования чисел
    if (typeof initNumberFormatting === 'function') {
        initNumberFormatting();
    }
}

/**
 * Настройка обработчиков событий
 */
function setupEventListeners() {
    // Кнопка расчёта
    const calculateBtn = document.getElementById('calculateBtn');
    if (calculateBtn) {
        calculateBtn.addEventListener('click', handleCalculate);
    }
    
    // Переключатель рассрочки
    const installmentToggle = document.getElementById('installmentToggle');
    if (installmentToggle) {
        installmentToggle.addEventListener('click', toggleInstallment);
    }
    
    // Переключатель единиц срока (месяцы/годы)
    const termToggles = document.querySelectorAll('.term-toggle .toggle-btn');
    termToggles.forEach(btn => {
        btn.addEventListener('click', function() {
            termToggles.forEach(b => b.classList.remove('active'));
            this.classList.add('active');
            termUnit = this.dataset.unit;
            updateTermInput();
        });
    });
    
    // Кнопка добавления досрочного платежа
    const addEarlyPaymentBtn = document.getElementById('addEarlyPaymentBtn');
    if (addEarlyPaymentBtn) {
        addEarlyPaymentBtn.addEventListener('click', showEarlyPaymentModal);
    }
    
    // Модальное окно досрочных платежей
    const modal = document.getElementById('earlyPaymentModal');
    const modalClose = document.querySelector('.modal-close');
    const cancelBtn = document.getElementById('cancelEarlyPaymentBtn');
    const saveBtn = document.getElementById('saveEarlyPaymentBtn');
    
    if (modalClose) {
        modalClose.addEventListener('click', hideEarlyPaymentModal);
    }
    if (cancelBtn) {
        cancelBtn.addEventListener('click', hideEarlyPaymentModal);
    }
    if (saveBtn) {
        saveBtn.addEventListener('click', saveEarlyPayment);
    }
    
    // Переключатель режима досрочного платежа
    const modeToggles = document.querySelectorAll('.mode-toggle .toggle-btn');
    modeToggles.forEach(btn => {
        btn.addEventListener('click', function() {
            modeToggles.forEach(b => b.classList.remove('active'));
            this.classList.add('active');
        });
    });
    
    // Кнопка экспорта в Excel
    const exportBtn = document.getElementById('exportBtn');
    if (exportBtn) {
        exportBtn.addEventListener('click', handleExport);
    }
    
    // Закрытие модального окна при клике вне его
    if (modal) {
        window.addEventListener('click', function(event) {
            if (event.target === modal) {
                hideEarlyPaymentModal();
            }
        });
    }
}

/**
 * Переключатель режима рассрочки
 */
function toggleInstallment() {
    const toggle = document.getElementById('installmentToggle');
    const rateGroup = document.getElementById('rateGroup');
    const rateInput = document.getElementById('rate');
    
    isInstallmentMode = !isInstallmentMode;
    
    if (isInstallmentMode) {
        toggle.classList.add('active');
        rateGroup.style.display = 'none';
        rateInput.value = '0';
    } else {
        toggle.classList.remove('active');
        rateGroup.style.display = 'block';
        if (rateInput.value === '0' || rateInput.value === '') {
            rateInput.value = '12';
        }
    }
}

/**
 * Обновление поля срока при переключении единиц
 */
function updateTermInput() {
    const termInput = document.getElementById('term');
    if (!termInput) return;
    
    let currentValue = parseInt(termInput.value) || 0;
    
    if (termUnit === 'years' && currentValue > 0) {
        // Конвертируем месяцы в годы
        termInput.value = Math.round(currentValue / 12);
    } else if (termUnit === 'months' && currentValue > 0) {
        // Конвертируем годы в месяцы
        termInput.value = currentValue * 12;
    }
}

/**
 * Основная функция расчёта кредита
 */
async function handleCalculate() {
    try {
        // Собираем данные из формы
        const principal = removeFormatting(document.getElementById('principal').value);
        const downPayment = removeFormatting(document.getElementById('downPayment').value) || 0;
        const termValue = parseInt(document.getElementById('term').value) || 0;
        const rate = parseFloat(document.getElementById('rate').value) || 0;
        const startDate = document.getElementById('startDate').value;
        
        // Валидация
        if (principal <= 0) {
            alert('Введите стоимость автомобиля');
            return;
        }
        
        if (termValue <= 0) {
            alert('Введите срок кредита');
            return;
        }
        
        // Конвертируем срок в месяцы
        let termMonths = termValue;
        if (termUnit === 'years') {
            termMonths = termValue * 12;
        }
        
        // Рассчитываем сумму кредита (стоимость - первоначальный взнос)
        const loanAmount = Math.max(0, principal - downPayment);
        
        if (loanAmount <= 0) {
            alert('Сумма кредита должна быть больше нуля');
            return;
        }
        
        // Подготавливаем данные для запроса
        const requestData = {
            principal: loanAmount,
            rate: isInstallmentMode ? 0 : rate,
            term_months: termMonths,
            start_date: startDate,
            early_payments: earlyPaymentsData
        };
        
        // Сохраняем параметры для экспорта
        currentParameters = {
            principal: principal,
            down_payment: downPayment,
            rate: isInstallmentMode ? 0 : rate,
            term_months: termMonths,
            start_date: startDate
        };
        
        // Отправляем запрос на сервер
        const response = await fetch('/api/calculate', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(requestData)
        });
        
        if (!response.ok) {
            throw new Error('Ошибка при расчёте кредита');
        }
        
        const result = await response.json();
        
        // Сохраняем расписание для экспорта
        currentSchedule = result.payment_schedule || [];
        
        // Отображаем результаты
        displayResults(result);
        
        // Обновляем диаграмму
        if (typeof createPaymentChart === 'function') {
            createPaymentChart({
                principal: result.principal,
                totalInterest: result.total_interest,
                totalEarlyPayment: result.total_early_payment
            });
        }
        
        // Обновляем таблицу
        updatePaymentTable(result.payment_schedule);
        
        // Активируем кнопку экспорта
        const exportBtn = document.getElementById('exportBtn');
        if (exportBtn) {
            exportBtn.disabled = false;
        }
        
    } catch (error) {
        console.error('Ошибка расчёта:', error);
        alert('Произошла ошибка при расчёте кредита. Проверьте введённые данные.');
    }
}

/**
 * Отображение результатов расчёта
 */
function displayResults(result) {
    // Сумма кредита
    const resultPrincipal = document.getElementById('resultPrincipal');
    if (resultPrincipal) {
        resultPrincipal.textContent = formatNumberDisplay(result.principal);
    }
    
    // Ежемесячный платёж
    const resultMonthlyPayment = document.getElementById('resultMonthlyPayment');
    if (resultMonthlyPayment) {
        resultMonthlyPayment.textContent = formatNumberDisplay(result.monthly_payment);
    }
    
    // Общая переплата
    const resultTotalInterest = document.getElementById('resultTotalInterest');
    if (resultTotalInterest) {
        resultTotalInterest.textContent = formatNumberDisplay(result.total_interest);
    }
    
    // Необходимый доход (платёж × 2.5)
    const resultRequiredIncome = document.getElementById('resultRequiredIncome');
    if (resultRequiredIncome) {
        const requiredIncome = result.monthly_payment * 2.5;
        resultRequiredIncome.textContent = formatNumberDisplay(requiredIncome);
    }
    
    // Итого к выплате
    const resultTotalAmount = document.getElementById('resultTotalAmount');
    if (resultTotalAmount) {
        resultTotalAmount.textContent = formatNumberDisplay(result.total_amount);
    }
    
    // Экономия (если есть)
    const savingsItem = document.getElementById('savingsItem');
    const resultSavings = document.getElementById('resultSavings');
    if (resultSavings && result.final_savings > 0) {
        resultSavings.textContent = formatNumberDisplay(result.final_savings);
        if (savingsItem) {
            savingsItem.style.display = 'flex';
        }
    } else if (savingsItem) {
        savingsItem.style.display = 'none';
    }
}

/**
 * Обновление таблицы платежей
 */
function updatePaymentTable(schedule) {
    const tbody = document.getElementById('paymentTableBody');
    const tfoot = document.getElementById('paymentTableFoot');
    
    if (!tbody) return;
    
    // Очищаем таблицу
    tbody.innerHTML = '';
    
    if (!schedule || schedule.length === 0) {
        tbody.innerHTML = '<tr><td colspan="7" class="empty-message">Нет данных для отображения</td></tr>';
        if (tfoot) tfoot.style.display = 'none';
        return;
    }
    
    // Подсчитываем итоги
    let totalMonthlyPayment = 0;
    let totalEarlyPayment = 0;
    let totalPrincipalPaid = 0;
    let totalInterestPaid = 0;
    
    // Заполняем таблицу
    schedule.forEach(payment => {
        const row = document.createElement('tr');
        
        totalMonthlyPayment += payment.monthly_payment || 0;
        totalEarlyPayment += payment.early_payment || 0;
        totalPrincipalPaid += payment.principal_paid || 0;
        totalInterestPaid += payment.interest_paid || 0;
        
        // Форматируем дату
        const paymentDate = new Date(payment.payment_date);
        const formattedDate = paymentDate.toLocaleDateString('ru-RU', {
            day: '2-digit',
            month: '2-digit',
            year: 'numeric'
        });
        
        row.innerHTML = `
            <td>${payment.month}</td>
            <td>${formattedDate}</td>
            <td>${formatNumberDisplay(payment.monthly_payment)}</td>
            <td>${formatNumberDisplay(payment.early_payment)}</td>
            <td>${formatNumberDisplay(payment.principal_paid)}</td>
            <td>${formatNumberDisplay(payment.interest_paid)}</td>
            <td>${formatNumberDisplay(payment.remaining_balance)}</td>
        `;
        
        tbody.appendChild(row);
    });
    
    // Обновляем итоговую строку
    if (tfoot) {
        document.getElementById('totalMonthlyPayment').textContent = formatNumberDisplay(totalMonthlyPayment);
        document.getElementById('totalEarlyPayment').textContent = formatNumberDisplay(totalEarlyPayment);
        document.getElementById('totalPrincipalPaid').textContent = formatNumberDisplay(totalPrincipalPaid);
        document.getElementById('totalInterestPaid').textContent = formatNumberDisplay(totalInterestPaid);
        tfoot.style.display = 'table-footer-group';
    }
}

/**
 * Показать модальное окно для добавления досрочного платежа
 */
function showEarlyPaymentModal() {
    const modal = document.getElementById('earlyPaymentModal');
    const monthSelect = document.getElementById('earlyMonth');
    
    if (!modal || !monthSelect) return;
    
    // Заполняем выпадающий список месяцев
    monthSelect.innerHTML = '';
    const maxMonths = parseInt(document.getElementById('term').value) || 36;
    const actualMaxMonths = termUnit === 'years' ? maxMonths * 12 : maxMonths;
    
    for (let i = 1; i <= actualMaxMonths; i++) {
        const option = document.createElement('option');
        option.value = i;
        option.textContent = `${i} месяц${i === 1 ? '' : i < 5 ? 'а' : 'ев'}`;
        monthSelect.appendChild(option);
    }
    
    // Очищаем поля
    document.getElementById('earlyAmount').value = '';
    const modeToggles = document.querySelectorAll('.mode-toggle .toggle-btn');
    modeToggles.forEach(btn => {
        btn.classList.remove('active');
        if (btn.dataset.mode === 'reduce_payment') {
            btn.classList.add('active');
        }
    });
    
    modal.style.display = 'block';
}

/**
 * Скрыть модальное окно
 */
function hideEarlyPaymentModal() {
    const modal = document.getElementById('earlyPaymentModal');
    if (modal) {
        modal.style.display = 'none';
    }
}

/**
 * Сохранить досрочный платёж
 */
function saveEarlyPayment() {
    const month = parseInt(document.getElementById('earlyMonth').value);
    const amount = removeFormatting(document.getElementById('earlyAmount').value);
    const activeModeBtn = document.querySelector('.mode-toggle .toggle-btn.active');
    const mode = activeModeBtn ? activeModeBtn.dataset.mode : 'reduce_payment';
    
    if (!month || month <= 0) {
        alert('Выберите месяц');
        return;
    }
    
    if (!amount || amount <= 0) {
        alert('Введите сумму досрочного платежа');
        return;
    }
    
    // Сохраняем досрочный платёж
    earlyPaymentsData[month] = {
        amount: amount,
        mode: mode
    };
    
    // Обновляем список досрочных платежей
    updateEarlyPaymentsList();
    
    // Закрываем модальное окно
    hideEarlyPaymentModal();
    
    // Автоматически пересчитываем, если уже был выполнен расчёт
    if (currentSchedule.length > 0) {
        handleCalculate();
    }
}

/**
 * Обновление списка досрочных платежей
 */
function updateEarlyPaymentsList() {
    const list = document.getElementById('earlyPaymentsList');
    if (!list) return;
    
    list.innerHTML = '';
    
    const months = Object.keys(earlyPaymentsData).sort((a, b) => parseInt(a) - parseInt(b));
    
    if (months.length === 0) {
        list.innerHTML = '<p style="color: var(--color-text-secondary); font-style: italic;">Досрочные платежи не добавлены</p>';
        return;
    }
    
    months.forEach(month => {
        const payment = earlyPaymentsData[month];
        const item = document.createElement('div');
        item.className = 'early-payment-item';
        
        const modeText = payment.mode === 'reduce_payment' ? 'Уменьшить платёж' : 'Уменьшить срок';
        
        item.innerHTML = `
            <div class="early-payment-info">
                <strong>Месяц ${month}: ${formatNumberDisplay(payment.amount)}</strong>
                <span>Режим: ${modeText}</span>
            </div>
            <button type="button" class="btn-remove" data-month="${month}">Удалить</button>
        `;
        
        // Обработчик удаления
        const removeBtn = item.querySelector('.btn-remove');
        removeBtn.addEventListener('click', function() {
            delete earlyPaymentsData[month];
            updateEarlyPaymentsList();
            
            // Автоматически пересчитываем, если уже был выполнен расчёт
            if (currentSchedule.length > 0) {
                handleCalculate();
            }
        });
        
        list.appendChild(item);
    });
}

/**
 * Экспорт в Excel
 */
async function handleExport() {
    try {
        if (!currentSchedule || currentSchedule.length === 0) {
            alert('Сначала выполните расчёт кредита');
            return;
        }
        
        const exportBtn = document.getElementById('exportBtn');
        if (exportBtn) {
            exportBtn.disabled = true;
            exportBtn.textContent = 'Экспорт...';
        }
        
        const requestData = {
            parameters: currentParameters,
            schedule: currentSchedule
        };
        
        const response = await fetch('/api/export', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(requestData)
        });
        
        if (!response.ok) {
            throw new Error('Ошибка при экспорте');
        }
        
        const result = await response.json();
        
        // Скачиваем файл
        window.location.href = result.url;
        
        if (exportBtn) {
            exportBtn.disabled = false;
            exportBtn.textContent = '⬇ Скачать таблицу в Excel';
        }
        
    } catch (error) {
        console.error('Ошибка экспорта:', error);
        alert('Произошла ошибка при экспорте в Excel');
        
        const exportBtn = document.getElementById('exportBtn');
        if (exportBtn) {
            exportBtn.disabled = false;
            exportBtn.textContent = '⬇ Скачать таблицу в Excel';
        }
    }
}

File: calculator.py
"""
Модуль для расчёта автокредитов с поддержкой досрочных платежей.
"""

from datetime import datetime, timedelta
from typing import Dict, List, Optional


def calculate_loan(
    principal: float,
    rate: float,
    term_months: int,
    start_date: Optional[str] = None,
    early_payments: Optional[Dict[int, Dict]] = None
) -> Dict:
    """
    Основной расчёт кредита.
    
    Args:
        principal (float): Сумма кредита (руб.)
        rate (float): Годовая процентная ставка (%)
        term_months (int): Срок кредита (месяцы)
        start_date (str): Дата получения кредита (YYYY-MM-DD)
        early_payments (dict): {месяц: {сумма, режим}} - досрочные платежи
    
    Returns:
        dict: {
            monthly_payment, total_interest, total_amount,
            payment_schedule, total_early_payment, final_savings
        }
    """
    if start_date is None:
        start_date = datetime.now().strftime('%Y-%m-%d')
    
    if early_payments is None:
        early_payments = {}
    
    # Месячная процентная ставка
    monthly_rate = rate / 100 / 12 if rate > 0 else 0
    
    # Расчёт аннуитетного платежа
    if monthly_rate > 0:
        monthly_payment = principal * (monthly_rate * (1 + monthly_rate) ** term_months) / \
                         ((1 + monthly_rate) ** term_months - 1)
    else:
        monthly_payment = principal / term_months
    
    # Генерация графика платежей
    schedule = generate_payment_schedule(
        principal, rate, term_months, start_date, early_payments
    )
    
    # Подсчёт итоговых значений
    total_interest = sum(payment['interest_paid'] for payment in schedule)
    total_early_payment = sum(
        payment.get('early_payment', 0) for payment in schedule
    )
    total_amount = principal + total_interest + total_early_payment
    
    # Расчёт экономии от досрочных платежей
    # Сравниваем с базовым расчётом без досрочных платежей
    if early_payments:
        base_schedule = generate_payment_schedule(
            principal, rate, term_months, start_date, None
        )
        base_total_interest = sum(p['interest_paid'] for p in base_schedule)
        base_total_amount = principal + base_total_interest
        final_savings = base_total_amount - total_amount
    else:
        final_savings = 0
    
    return {
        'monthly_payment': monthly_payment,
        'total_interest': total_interest,
        'total_amount': total_amount,
        'payment_schedule': schedule,
        'total_early_payment': total_early_payment,
        'final_savings': final_savings,
        'principal': principal
    }


def generate_payment_schedule(
    principal: float,
    rate: float,
    term_months: int,
    start_date: str,
    early_payments: Optional[Dict[int, Dict]] = None
) -> List[Dict]:
    """
    Генерирует детальный график платежей.
    
    Returns:
        list[dict]: Каждый элемент:
            {
                month, payment_date, monthly_payment, early_payment,
                principal_paid, interest_paid, remaining_balance
            }
    """
    if early_payments is None:
        early_payments = {}
    
    monthly_rate = rate / 100 / 12 if rate > 0 else 0
    
    # Базовая сумма аннуитетного платежа (без досрочных платежей)
    if monthly_rate > 0:
        base_monthly_payment = principal * (monthly_rate * (1 + monthly_rate) ** term_months) / \
                              ((1 + monthly_rate) ** term_months - 1)
    else:
        base_monthly_payment = principal / term_months
    
    schedule = []
    remaining_balance = principal
    start = datetime.strptime(start_date, '%Y-%m-%d')
    current_month = 1
    current_monthly_payment = base_monthly_payment
    remaining_term = term_months
    
    while remaining_balance > 0.01 and current_month <= term_months * 3:  # Защита от бесконечного цикла
        # Дата платежа
        payment_date = start + timedelta(days=30 * (current_month - 1))
        
        # Проценты за месяц
        interest_paid = remaining_balance * monthly_rate
        
        # Основной платёж из ежемесячного платежа
        if monthly_rate > 0:
            principal_from_payment = current_monthly_payment - interest_paid
        else:
            principal_from_payment = current_monthly_payment
        
        # Досрочный платёж
        early_payment = 0
        early_payment_mode = None
        
        if current_month in early_payments:
            early_payment = early_payments[current_month]['amount']
            early_payment_mode = early_payments[current_month].get('mode', 'reduce_payment')
        
        # Общая сумма, идущая на погашение основного долга
        total_principal_paid = principal_from_payment
        
        if early_payment > 0:
            if early_payment_mode == 'reduce_term':
                # Режим "уменьшить срок": досрочный платёж полностью идёт на основной долг
                total_principal_paid += early_payment
            else:
                # Режим "уменьшить платёж": досрочный платёж идёт на основной долг,
                # но ежемесячный платёж остаётся прежним
                total_principal_paid += early_payment
        
        # Проверка на переплату
        if total_principal_paid > remaining_balance:
            total_principal_paid = remaining_balance
        
        remaining_balance -= total_principal_paid
        
        # Если остаток стал отрицательным или очень маленьким
        if remaining_balance < 0.01:
            remaining_balance = 0
        
        schedule.append({
            'month': current_month,
            'payment_date': payment_date.strftime('%Y-%m-%d'),
            'monthly_payment': current_monthly_payment,
            'early_payment': early_payment,
            'principal_paid': total_principal_paid,
            'interest_paid': interest_paid,
            'remaining_balance': remaining_balance
        })
        
        # Если остаток погашен, выходим
        if remaining_balance <= 0.01:
            break
        
        current_month += 1
        remaining_term -= 1
        
        # Пересчёт ежемесячного платежа при досрочном платеже с уменьшением срока
        if early_payment > 0 and early_payment_mode == 'reduce_term' and remaining_balance > 0.01:
            # Пересчитываем аннуитетный платёж с учётом нового остатка и оставшегося срока
            if remaining_term > 0:
                if monthly_rate > 0:
                    current_monthly_payment = remaining_balance * \
                        (monthly_rate * (1 + monthly_rate) ** remaining_term) / \
                        ((1 + monthly_rate) ** remaining_term - 1)
                else:
                    current_monthly_payment = remaining_balance / remaining_term
            else:
                current_monthly_payment = remaining_balance
    
    return schedule


def apply_early_payment(
    schedule: List[Dict],
    month: int,
    amount: float,
    mode: str = 'reduce_payment'
) -> List[Dict]:
    """
    Применяет досрочный платёж и пересчитывает график.
    
    Args:
        schedule: Текущий график платежей
        month: Номер месяца для досрочного платежа
        amount: Сумма досрочного платежа
        mode: 'reduce_payment' (уменьшить платёж) или 'reduce_term' (уменьшить срок)
    
    Returns:
        list[dict]: Обновлённый график платежей
    """
    # Эта функция используется для интерактивного пересчёта
    # В основном расчёте досрочные платежи обрабатываются в generate_payment_schedule
    return schedule
